### PR TITLE
filter out tests waiting for next tfp release

### PR DESCRIPTION
**Pull request by juanitorduz**
Partially addresses https://github.com/pyro-ppl/numpyro/issues/1814 . We must keep in mind removing these `skip` statements once we see a new release.

### CODE REVIEW

**Comment by juanitorduz** [comment]
There is a new test failing probably because the new jax and/or numpy releases

```python
__________________ test_discrete_site_without_infer_enumerate __________________

    def test_discrete_site_without_infer_enumerate():
        def model():
            numpyro.sample("x", dist.Bernoulli(0.5))
    
        mcmc = MCMC(NUTS(model), num_warmup=10, num_samples=10)
        with pytest.warns(FutureWarning, match="enumerated sites"):
>           mcmc.run(random.PRNGKey(0))
E           FutureWarning: unhashable type: <class 'jax._src.interpreters.partial_eval.DynamicJaxprTracer'>. Attempting to hash a tracer will lead to an error in a future JAX release.

test/infer/test_mcmc.py:1104: FutureWarning
```

If I remove the `warn` match, it will pass. But I have not been able to figure out which match to pass as there are many coming from various places now 😅

**Comment by juanitorduz** [comment]
It seems 

```pyhton
FutureWarning: unhashable type: <class 'jax._src.interpreters.batching.BatchTracer'>. Attempting to hash a tracer will lead to an error in a future JAX release.
```

is all over the place now 😓

**Comment by fehiepsi** [comment]
Thanks @juanitorduz. I'm looking at them.

**Comment by juanitorduz** [comment]
> Thanks @juanitorduz. I'm looking at them.

ok! You can either push to this branch or create a new one if needed

**Comment by fehiepsi** [comment]
It turns out that in funsor, we have some checks for tracers to be [Hashable](https://github.com/pyro-ppl/funsor/blob/c96e70c89aaa0bda8dc132ae1c3498a681f618b8/funsor/interpretations.py#L62). I don't think that the new behavior will cause issues: it is fine to let arrays to be either hashable or unhashable. So I think we can simply filter out these warnings:
+ in `__init__`, at the top of the file:
```
import warnings

warnings.filterwarnings("ignore", message=".*Attempting to hash a tracer.*", category=FutureWarning)
```
+ in `pyproject.toml`: `"ignore:.*Attempting to hash a tracer:FutureWarning",` for pytest

**Comment by juanitorduz** [comment]
ok! We are making progress 😅 ! Now we have 

```python
FAILED test/contrib/test_control_flow.py::test_scan - TypeError: body_fun output and input must have identical types, got
('ShapedArray(int32[], weak_type=True)', ['ShapedArray(float32[10])', 'ShapedArray(float32[10])', 'DIFFERENT ShapedArray(int32[], weak_type=True) vs. ShapedArray(float0[])', 'ShapedArray(float32[])', 'ShapedArray(float32[])'], []).
FAILED test/contrib/test_control_flow.py::test_scan_svi - TypeError: body_fun output and input must have identical types, got
('ShapedArray(int32[], weak_type=True)', ['ShapedArray(float32[3,5])', 'DIFFERENT ShapedArray(int32[], weak_type=True) vs. ShapedArray(float0[])', 'ShapedArray(float32[5])'], []).
```

and 

```python
TypeError: body_fun output and input must have identical types, got
('ShapedArray(int32[], weak_type=True)', ['ShapedArray(float32[])', 'ShapedArray(float32[])', 'ShapedArray(float32[])', 'ShapedArray(float32[])', 'DIFFERENT ShapedArray(int32[], weak_type=True) vs. ShapedArray(float0[])', 'ShapedArray(float32[])', 'ShapedArray(float32[])', 'ShapedArray(float32[10])'], []).
test/test_examples.py::test_cpu[holt_winters.py --T 4 --num-samples 10 --num-warmup 10 --num-chains 2] Running:
python examples/holt_winters.py --T 4 --num-samples 10 --num-warmup 10 --num-chains 2
```

I am so seeing 

```python
The preceding stack trace is the source of the JAX operation that, once transformed by JAX, triggered the following exception.

--------------------

The above exception was the direct cause of the following exception:

jax.errors.SimplifiedTraceback: For simplicity, JAX has removed its internal frames from the traceback of the following exception. Set JAX_TRACEBACK_FILTERING=off to include these.
```

🤔

**Comment by juanitorduz** [comment]
It seems like a type problem ... could this be again a numpy or jax recent change?

File: numpyro/infer/__init__.py [hunk 1]
```diff
@@ -29,6 +31,11 @@
 
 from . import autoguide, reparam
 
+warnings.filterwarnings(
+    "ignore", message=".*Attempting to hash a tracer.*", category=FutureWarning
+)
```
Comment (fehiepsi) [inline numpyro/infer/__init__.py:36]: Could we move this to `__init__` before all imports? I found that the warning still appears in some examples like `examples/gaussian_shells.py --enum`.

Comment (juanitorduz) [reply numpyro/infer/__init__.py]: Changed in https://github.com/pyro-ppl/numpyro/pull/1817/commits/0afb74689a04acebdad100027115586d800a3567

Comment (juanitorduz) [reply numpyro/infer/__init__.py]: But I think this does not fix the type errors right?

Comment (fehiepsi) [reply numpyro/infer/__init__.py]: Do you think that moving this to `numpyro/__init__.py` is better? There are other places that we import `funsor`.

Comment (fehiepsi) [reply numpyro/infer/__init__.py]: Re type error: let me take further look. I guess we just need to cast types properly.

Comment (juanitorduz) [reply numpyro/infer/__init__.py]: I tried but have not been able to fix the types :(

**Comment by fehiepsi** [comment]
Hi @juanitorduz, I raised the upstream error in https://github.com/google/jax/issues/22045. For a fix, could you help me change every `device_put(foo)` in https://github.com/pyro-ppl/numpyro/blob/master/numpyro/contrib/control_flow/scan.py to
```python
tree_map(device_put, foo)
```

**Comment by juanitorduz** [comment]
In https://github.com/pyro-ppl/numpyro/pull/1817/commits/19f82322060fd1b6ce31be1797f98b254610c2bf I still saw other test failing:

```python
FAILED test/contrib/test_tfp.py::test_unnormalized_normal_chain[1-MetropolisAdjustedLangevinAlgorithm-kwargs0] - OverflowError: Python int too large to convert to C long
FAILED test/contrib/test_tfp.py::test_unnormalized_normal_chain[1-SliceSampler-kwargs2] - OverflowError: Python int too large to convert to C long
FAILED test/contrib/test_tfp.py::test_unnormalized_normal_chain[1-UncalibratedLangevin-kwargs3] - OverflowError: Python int too large to convert to C long
FAILED test/contrib/test_tfp.py::test_unnormalized_normal_chain[2-MetropolisAdjustedLangevinAlgorithm-kwargs0] - OverflowError: Python int too large to convert to C long
FAILED test/contrib/test_tfp.py::test_unnormalized_normal_chain[2-SliceSampler-kwargs2] - OverflowError: Python int too large to convert to C long
FAILED test/contrib/test_tfp.py::test_unnormalized_normal_chain[2-UncalibratedLangevin-kwargs3] - OverflowError: Python int too large to convert to C long
```

Hence: https://github.com/pyro-ppl/numpyro/pull/1817/commits/038dd84ad2ad51ba2d0ae8f8add1f3893c16db63

**Comment by juanitorduz** [comment]
Ok! Finally is 🟢! 

Shall we revert these changes once the JAX issue is fixed and released? Similarly for tfp next release?

**Review by fehiepsi — APPROVED** [review]
Thanks for addressing the issues, @juanitorduz!!

**Pull request merged**